compiler panic with `cut(.)`

```
11:31 ~/work/brim/zq(shaping-functions)
$ zq -t "put a=cut(.)" all.zng 

panic: runtime error: slice bounds out of range [:-1]

goroutine 1 [running]:
github.com/brimsec/zq/zng/builder.NewColumnBuilder(0xc00028df40, 0xc0003a80a0, 0x1, 0x1, 0xf, 0x0, 0x0)
        /Users/henridf/work/brim/zq/zng/builder/colbuilder.go:97 +0x9ba
github.com/brimsec/zq/expr.NewCutter(0xc00028df40, 0xc0003a80a0, 0x1, 0x1, 0xc0002e80a0, 0x1, 0x1, 0xf, 0x0, 0x0)
        /Users/henridf/work/brim/zq/expr/cutter.go:33 +0x50
github.com/brimsec/zq/compiler.compileCutter(0xc00028df40, 0x18ca115, 0xc, 0xc0002d33c0, 0x3, 0xc0002e8f50, 0x1, 0x1, 0x170, 0x8a4087ac145c319c, ...)
        /Users/henridf/work/brim/zq/compiler/expr.go:261 +0x388
github.com/brimsec/zq/compiler.compileCall(0xc00028df40, 0x18ca115, 0xc, 0xc0002d33c0, 0x3, 0xc0002e8f50, 0x1, 0x1, 0x2d1d600, 0x20300000000000, ...)
        /Users/henridf/work/brim/zq/compiler/expr.go:292 +0x6f1
github.com/brimsec/zq/compiler.CompileExpr(0xc00028df40, 0x1a034e0, 0xc00028dd80, 0x28, 0x28, 0x27867d0, 0xc0002f00f0)
        /Users/henridf/work/brim/zq/compiler/expr.go:59 +0x547
github.com/brimsec/zq/compiler.CompileAssignment(0xc00028df40, 0xc00079f228, 0x0, 0x0, 0x0, 0x0, 0x0, 0x100, 0x2efffff)
        /Users/henridf/work/brim/zq/compiler/expr.go:194 +0x79
github.com/brimsec/zq/compiler.compileAssignments(0xc0002f04b0, 0x1, 0x1, 0xc00028df40, 0x203000, 0x8, 0x2000, 0x203000, 0x203000)
        /Users/henridf/work/brim/zq/compiler/proc.go:210 +0x127
github.com/brimsec/zq/compiler.compileProc(0x0, 0x1a03700, 0xc0002f04e0, 0xc0002f00c0, 0x1a0bf40, 0xc0002e8090, 0xa, 0x1a03380, 0xc00028dd40, 0x1a034e0)
        /Users/henridf/work/brim/zq/compiler/proc.go:146 +0x657
github.com/brimsec/zq/compiler.Compile(0x0, 0x1a03700, 0xc0002f04e0, 0xc0002f00c0, 0xc0002e8070, 0x1, 0x1, 0x0, 0x100000002, 0x203000, ...)
        /Users/henridf/work/brim/zq/compiler/proc.go:333 +0x585
github.com/brimsec/zq/compiler.compileSequential(0x0, 0xc0002e8f30, 0x1, 0x1, 0xc0002f00c0, 0xc0002e8070, 0x1, 0x1, 0x100f690, 0xc0002f00c0, ...)
        /Users/henridf/work/brim/zq/compiler/proc.go:240 +0xb4
github.com/brimsec/zq/compiler.Compile(0x0, 0x1a037c0, 0xc0002f0510, 0xc0002f00c0, 0xc0002e8070, 0x1, 0x1, 0x7fffffffffffffff, 0x1a0c900, 0xc00029d0e0, ...)
        /Users/henridf/work/brim/zq/compiler/proc.go:301 +0x737
github.com/brimsec/zq/driver.compile(0x1a14140, 0xc000394100, 0x1a037c0, 0xc0002f0510, 0xc00028df40, 0xc00079fac8, 0x1, 0x1, 0x0, 0xc00010e240, ...)
        /Users/henridf/work/brim/zq/driver/compile.go:77 +0x545
github.com/brimsec/zq/driver.Run(0x1a14700, 0xc000710060, 0x1a14440, 0xc0002f0060, 0x1a037c0, 0xc0002f0510, 0xc00028df40, 0x1a04100, 0xc000710030, 0x0, ...)
        /Users/henridf/work/brim/zq/driver/driver.go:29 +0x145
main.(*Command).Run(0xc000296c00, 0xc0000320e0, 0x2, 0x2, 0x0, 0x0)
        /Users/henridf/work/brim/zq/cmd/zq/zq.go:180 +0x9b5
github.com/mccanne/charm.(*instance).run(0xc00000fc20, 0xc0000320d0, 0x3, 0x3, 0x0, 0x0)
        /Users/henridf/go/pkg/mod/github.com/mccanne/charm@v0.0.3-0.20191224190439-b05e1b7b1be3/instance.go:53 +0x293
github.com/mccanne/charm.(*Spec).ExecRoot(0x1e53d00, 0xc0000320d0, 0x3, 0x3, 0xc00019ff78, 0x1007f65, 0xc00008e058, 0x0)
        /Users/henridf/go/pkg/mod/github.com/mccanne/charm@v0.0.3-0.20191224190439-b05e1b7b1be3/charm.go:77 +0x96
main.main()
        /Users/henridf/work/brim/zq/cmd/zq/main.go:9 +0x85
```

Similarly when using `cut` as a proc:

```
$ zq -t "cut .=id" input.tzng 
panic: runtime error: slice bounds out of range [:-1]

goroutine 1 [running]:
github.com/brimsec/zq/zng/builder.NewColumnBuilder(0xc0002cb100, 0xc00030e080, 0x1, 0x1, 0x1a094c0, 0xc00030e060, 0x23c09f0)
	/Users/henridf/work/brim/zq/zng/builder/colbuilder.go:97 +0x9ba
github.com/brimsec/zq/expr.NewCutter(0xc0002cb100, 0xc00030e080, 0x1, 0x1, 0xc0003100a0, 0x1, 0x1, 0x1, 0x1, 0x10)
	/Users/henridf/work/brim/zq/expr/cutter.go:33 +0x50
github.com/brimsec/zq/compiler.compileProc(0x0, 0x1a08dc0, 0xc0002e7b90, 0xc0003120c0, 0x1a11820, 0xc000310090, 0x400, 0x2c18e00, 0x20300000000000, 0x2dfffff)
	/Users/henridf/work/brim/zq/compiler/proc.go:66 +0x234d
github.com/brimsec/zq/compiler.Compile(0x0, 0x1a08dc0, 0xc0002e7b90, 0xc0003120c0, 0xc000310070, 0x1, 0x1, 0x0, 0x0, 0x64, ...)
	/Users/henridf/work/brim/zq/compiler/proc.go:333 +0x585
github.com/brimsec/zq/compiler.compileSequential(0x0, 0xc0002db6c0, 0x1, 0x1, 0xc0003120c0, 0xc000310070, 0x1, 0x1, 0x100f690, 0xc0003120c0, ...)
	/Users/henridf/work/brim/zq/compiler/proc.go:240 +0xb4
github.com/brimsec/zq/compiler.Compile(0x0, 0x1a09140, 0xc0002e7bc0, 0xc0003120c0, 0xc000310070, 0x1, 0x1, 0x7fffffffffffffff, 0x1a11f60, 0xc0003080c0, ...)
	/Users/henridf/work/brim/zq/compiler/proc.go:301 +0x737
github.com/brimsec/zq/driver.compile(0x1a19a20, 0xc000314000, 0x1a09140, 0xc0002e7bc0, 0xc0002cb100, 0xc00021fac8, 0x1, 0x1, 0x0, 0xc000308060, ...)
	/Users/henridf/work/brim/zq/driver/compile.go:77 +0x545
github.com/brimsec/zq/driver.Run(0x1a19fe0, 0xc0002e7ec0, 0x1a19d20, 0xc000312060, 0x1a09140, 0xc0002e7bc0, 0xc0002cb100, 0x1a099e0, 0xc0002e7e90, 0x0, ...)
	/Users/henridf/work/brim/zq/driver/driver.go:29 +0x145
main.(*Command).Run(0xc000103980, 0xc000138020, 0x2, 0x2, 0x0, 0x0)
	/Users/henridf/work/brim/zq/cmd/zq/zq.go:180 +0x9b5
github.com/mccanne/charm.(*instance).run(0xc000129ca0, 0xc000138010, 0x3, 0x3, 0x0, 0x0)
	/Users/henridf/go/pkg/mod/github.com/mccanne/charm@v0.0.3-0.20191224190439-b05e1b7b1be3/instance.go:53 +0x293
github.com/mccanne/charm.(*Spec).ExecRoot(0x1e5ad20, 0xc000138010, 0x3, 0x3, 0xc00021ff78, 0x1007f65, 0xc00010e058, 0x0)
	/Users/henridf/go/pkg/mod/github.com/mccanne/charm@v0.0.3-0.20191224190439-b05e1b7b1be3/charm.go:77 +0x96
main.main()
	/Users/henridf/work/brim/zq/cmd/zq/main.go:9 +0x85
```

--- a/expr/cutter.go
+++ b/expr/cutter.go
@@ -40,7 +40,7 @@ func NewCutter(zctx *resolver.Context, fieldRefs []field.Static, fieldExprs []Ev
 		}
 	}
 	var b *builder.ColumnBuilder
-	if len(fieldRefs) > 0 && !fieldRefs[0].IsRoot() {
+	if len(fieldRefs) == 0 || !fieldRefs[0].IsRoot() {
 		// A root field will cause NewColumnBuilder to panic.
 		var err error
 		b, err = builder.NewColumnBuilder(zctx, fieldRefs)

--- a/expr/cutter.go
+++ b/expr/cutter.go
@@ -1,7 +1,9 @@
 package expr
 
 import (
+	"errors"
 	"fmt"
+	"github.com/brimsec/zq/zcode"
 	"strings"
 
 	"github.com/brimsec/zq/field"
@@ -30,9 +32,21 @@ type Cutter struct {
 // is false, the Cutter copies any fields in fieldnames, where targets
 // specifies the copied field names.
 func NewCutter(zctx *resolver.Context, fieldRefs []field.Static, fieldExprs []Evaluator) (*Cutter, error) {
-	b, err := builder.NewColumnBuilder(zctx, fieldRefs)
-	if err != nil {
-		return nil, err
+	if len(fieldRefs) > 1 {
+		for _, f := range fieldRefs {
+			if f.IsRoot() {
+				return nil, errors.New("cannot assign to . when cutting multiple values")
+			}
+		}
+	}
+	var b *builder.ColumnBuilder
+	if len(fieldRefs) > 0 && !fieldRefs[0].IsRoot() {
+		// A root field will cause NewColumnBuilder to panic.
+		var err error
+		b, err = builder.NewColumnBuilder(zctx, fieldRefs)
+		if err != nil {
+			return nil, err
+		}
 	}
 	return &Cutter{
 		zctx:        zctx,
@@ -59,6 +73,18 @@ func (c *Cutter) FoundCut() bool {
 // receiver's configuration.  If the resulting record would be empty, Apply
 // returns nil.
 func (c *Cutter) Apply(in *zng.Record) (*zng.Record, error) {
+	if len(c.fieldRefs) == 1 && c.fieldRefs[0].IsRoot() {
+		zv, err := c.fieldExprs[0].Eval(in)
+		if err != nil {
+			return nil, err
+		}
+		recType, ok := zng.AliasedType(zv.Type).(*zng.TypeRecord)
+		if !ok {
+			return nil, errors.New("cannot cut a non-record to .")
+		}
+		c.dirty = true
+		return zng.NewRecord(recType, append(zcode.Bytes{}, zv.Bytes...)), nil
+	}
 	types := c.typeCache
 	b := c.builder
 	b.Reset()

--- a/expr/ztests/cut-root-error.yaml
+++ b/expr/ztests/cut-root-error.yaml
@@ -1,0 +1,7 @@
+zql: 'put x=cut(.,b)'
+
+input: |
+  #0:record[a:int32,b:int32]
+  0:[1;2;]
+
+errorRE: cannot assign to \. when cutting multiple values

--- a/expr/ztests/cut-root.yaml
+++ b/expr/ztests/cut-root.yaml
@@ -1,0 +1,13 @@
+zql: 'put x=cut(.)'
+
+input: |
+  #0:record[a:record[b:int32]]
+  0:[-;]
+  0:[[-;]]
+  0:[[1;]]
+
+output: |
+  #0:record[a:record[b:int32],x:record[a:record[b:int32]]]
+  0:[-;[-;]]
+  0:[[-;][[-;]]]
+  0:[[1;][[1;]]]

--- a/expr/ztests/cut-to-root-error-multiple-values.yaml
+++ b/expr/ztests/cut-to-root-error-multiple-values.yaml
@@ -1,0 +1,7 @@
+zql: 'cut .=a,x=b'
+
+input: |
+  #0:record[a:int32,b:int32]
+  0:[1;2;]
+
+errorRE: cannot assign to \. when cutting multiple values

--- a/expr/ztests/cut-to-root-error-non-record.yaml
+++ b/expr/ztests/cut-to-root-error-non-record.yaml
@@ -1,0 +1,7 @@
+zql: 'cut .=a'
+
+input: |
+  #0:record[a:int32]
+  0:[1;]
+
+errorRE: cannot cut a non-record to \.

--- a/expr/ztests/cut-to-root.yaml
+++ b/expr/ztests/cut-to-root.yaml
@@ -1,0 +1,13 @@
+zql: 'cut .=a'
+
+input: |
+  #0:record[a:record[b:int32]]
+  0:[-;]
+  0:[[-;]]
+  0:[[1;]]
+
+output: |
+  #0:record[b:int32]
+  0:[]
+  0:[-;]
+  0:[1;]

